### DISCUSSION TITLE
/source/drivers/network/main.c

### DISCUSSION THREAD

**Discussion by defdefred** [VendelinSlezak/BleskOS#37] [Ideas]
Looking at the way you are choosing the right driver function for each network card.
```
void network_send_packet(dword_t memory, dword_t length) {
 if(ethernet_cards[ethernet_selected_card].driver==NETWORK_DRIVER_ETHERNET_INTEL_E1000) {
   ethernet_card_intel_e1000_send_packet(memory, length);
 }
 else if(ethernet_cards[ethernet_selected_card].driver==NETWORK_DRIVER_ETHERNET_AMD_PCNET) {
   ethernet_card_amd_pcnet_send_packet(memory, length);
 }
 else if(ethernet_cards[ethernet_selected_card].driver==NETWORK_DRIVER_ETHERNET_REALTEK_8139) {
   ethernet_card_realtek_8139_send_packet(memory, length);
 }
 else if(ethernet_cards[ethernet_selected_card].driver==NETWORK_DRIVER_ETHERNET_REALTEK_8169) {
   ethernet_card_realtek_8169_send_packet(memory, length);
 }
}
```
Seems that the use of function pointer could help to set the rigth function one time for all.
https://stackoverflow.com/questions/840501/how-do-function-pointers-in-c-work

**Comment by defdefred** [comment]
Same with `drivers/graphic/main.c`, instead of testing all the time if the frame buffer is 24bpp or 32bpp, just create separate functions and configure the good function pointer for redraw at graphic initialization.

Also, i can see that supporting 16bpp framebuffer could be easy :-)

**Comment by VendelinSlezak** [comment]
Thank you for idea, I will surely give it look and apply those things maybe already in Update 2.

**Comment by VendelinSlezak** [comment]
I added all of this in Update 2.
